### ISSUE TITLE
`missing_copy_implementations` shouldn't warn for unions

### ISSUE THREAD

**Issue by asomers** [rust-lang/rust#115014]
### Code

```Rust
#![deny(missing_copy_implementations)]
#![allow(dead_code)]

pub union Foo {
    int: i16,
    float: f64
}

pub struct Bar {
    discriminant: i8,
    foo: Foo
}
```

### Current output

```Shell
error: type could implement `Copy`; consider adding `impl Copy`
 --> src/lib.rs:3:1
  |
3 | / pub union Foo {
4 | |     int: i16,
5 | |     float: f64
6 | | }
  | |_^
  |
note: the lint level is defined here
 --> src/lib.rs:1:9
  |
1 | #![deny(missing_copy_implementations)]
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
```


### Desired output

```Shell
No error should be reported.
```

### Rationale and extra context

A union type might not be fully initialized.  For example, `Foo`'s fields are differently sized.  Initializing like `Foo{int: 0}` would leave 6 bytes uninitialized.  Therefore, trying to copy it byte-by-byte could read uninitialized fields, which is UB.  In this example of typical usage, the `Bar` struct contains a discriminant field which identifies which field of the enclosed union is valid.  So `Bar` can implement `Clone` even though it can't implement `Copy`.  But `Foo` by itself cannot safely implement either.

The compiler should not warn that a union like Foo is missing a `copy` implementation.

**Comment by asomers** [comment]
Also, the same bug applies for `missing_debug_implementations`.

**Comment by asquared31415** [comment]
`Foo` can safely implement `Copy` and `Clone`, in fact the `derive` macros work. I believe that trying to copy a `union` doesn't try to read *any* of the bytes as initialized, so it doesn't matter.

`Debug` is a different story, the `derive` is correctly rejected, and there's not a reasonable way to implement it for a union (besides perhaps something like `UnsafeCell` where nothing is emitted but the type name).

**Comment by asomers** [comment]
If copying a union doesn't read *any* of the bytes as initialized, then how can it possibly work?  Or are you saying that this is an exception to the "never read uninitialized data" rule?

**Comment by ShE3py** [comment]
https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html

> Uninitialized memory is also implicitly invalid for any type that has a restricted set of valid values. In other words, the only cases in which reading uninitialized memory is permitted are inside `union`s and in "padding" (the gaps between the fields/elements of a type).

I think UB occurs when trying to interpret uninitialized memory as a `T` (e.g. `u32`), `memcpy` is fine since we're just copying bits from A to B, and if it was ok to store A's bits, it should be ok to store them twice.

Furthermore all types should implement `Debug`, even if it's not meaningful :
https://github.com/rust-lang/rust/blob/fe5f5912579d02c4e6088c89574493dfa988531a/library/core/src/mem/maybe_uninit.rs#L263-L268

**Comment by asomers** [comment]
Interesting.  So there _is_ a specific exception for reading uninitialized unions.  Thanks for sharing that.  I guess this bug isn't valid then.
